lib/Db/MessageMapper.php

```diff
@@ -429,6 +431,25 @@ public function updateBulk(Account $account, bool $permflagsEnabled, Message ...
 		return $messages;
 	}
 
+	/**
+	 * @param int[] $uids
+	 */
+	public function updateFlagBulk(Mailbox $mailbox, string $flag, bool $set, array $uids): void {
+		$col = $this->hordeFlagToColumnName($flag);
+		$query = $this->db->getQueryBuilder();
+		$query->update($this->getTableName())
+			->set($col, $query->createNamedParameter($set, IQueryBuilder::PARAM_BOOL))
+			->where(
+				$query->expr()->eq('mailbox_id', $query->createNamedParameter($mailbox->getId(), IQueryBuilder::PARAM_INT), IQueryBuilder::PARAM_INT),
+				$query->expr()->eq($col, $query->createNamedParameter(!$set, IQueryBuilder::PARAM_BOOL), IQueryBuilder::PARAM_BOOL),
+				$query->expr()->in('uid', $query->createParameter('uids'), IQueryBuilder::PARAM_INT_ARRAY),
```

Comment:
I'm not sure if possible with all your fancy pancy supported databases

but an index like

`('mailbox_id', 'uid') INCLUDING (flag1, flag2, flag3, flag4, flag5, flag6)` would allow the filtering to happen mainly on the index level. Without going to the actual data.